### PR TITLE
Remove `GetDataDir(net_specific)` function

### PR DESCRIPTION
**Pull request by kiminuo**
This PR is a follow up PR to #21244. The PR attempts to move us an inch towards the [goal](https://github.com/bitcoin/bitcoin/pull/21244#discussion_r615307465) by removing `GetDataDir(net_specific)` and replacing it by `gArgs.GetDataDir(net_specific)` calls. 

The approach of this PR attempts to be similar to the one chosen in "De-globalize ChainstateManager" (#20158). The goal is to pass `ArgsManager` to functions (or ideally to have `ArgsManager` as a member of a class where needed; inspiration from here: #21789) instead of having it as a global variable (i.e. `gArgs`).

**Notes:**
* First commit makes `m_cached_blocks_path` `mutable` as was suggested [here](https://github.com/bitcoin/bitcoin/pull/21244#discussion_r615274095) but not fully applied in #21244. (`m_cached_datadir_path` and `m_cached_network_datadir_path` were marked as `mutable` in #21244) This commit can be in a separate PR too.
* Other commits deal with removing of `GetDataDir(net_specific)` function.
    * This was originally part of #21244 but it was [left]((https://github.com/bitcoin/bitcoin/pull/21244#pullrequestreview-633779754)) for a follow up PR.
* I think that the proposed changes show nicely where there is reliance on `gArgs` which is IMO a good thing.

If you know about a better approach how to do this, please share it here.

### CODE REVIEW

**Comment by laanwj** [comment]
Code review and concept ACK 2ca79de5a92be6f5179c47a4e28bcb5109a3670b
Getting the data directory seems common enough (as well as shared between all commands) to have on gArgs directly.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 2ca79de5a92be6f5179c47a4e28bcb5109a3670b.

Just my opinion, but I do think since this is switching code to a new ArgsManager interface, it would be better for the interface to have explicit Base and Net methods instead of using the previous `bool net_specific=true` default argument. E.g. something like:

```c++
public:
    const fs::path& GetDataDirBasePath() const { return GetDataDirPath(false); }
    const fs::path& GetDataDirNetPath() const { return GetDataDirPath(true); }
private:
    const fs::path& GetDataDirPath(bool net_specific) const;
```

`GetDataDirNetPath()` would seem more readable than `GetDataDirPath(true)` at call sites, and harder to confuse when writing code. And since call sites are being updated in this PR anyway, implementing this would just require tweaking the scripted diffs.

**Comment by kiminuo** [comment]
> Just my opinion, but I do think since this is switching code to a new ArgsManager interface, it would be better for the interface to have explicit Base and Net methods instead of using the previous `bool net_specific=true` default argument. E.g. something like:

I find this to be a very good suggestion. If this PR needs to be touched, I would do it in this PR. Otherwise, I would reserve that for a follow-up PR.

**Comment by MarcoFalke** [comment]
Is there a reason to suffix the return type to the function name? Wouldn't `GetDataDirBase` and `GetDataDirNet` be sufficient?

**Comment by laanwj** [comment]
> GetDataDirNetPath() would seem more readable than GetDataDirPath(true) at call sites,

Agree. Boolean arguments tend to be bad for readability.

**Comment by kiminuo** [comment]
> 
> 
> Code review ACK [2ca79de](https://github.com/bitcoin/bitcoin/commit/2ca79de5a92be6f5179c47a4e28bcb5109a3670b).
> 
> Just my opinion, but I do think since this is switching code to a new ArgsManager interface, it would be better for the interface to have explicit Base and Net methods instead of using the previous `bool net_specific=true` default argument. E.g. something like:
> 
> ```c++
> public:
>     const fs::path& GetDataDirBasePath() const { return GetDataDirPath(false); }
>     const fs::path& GetDataDirNetPath() const { return GetDataDirPath(true); }
> private:
>     const fs::path& GetDataDirPath(bool net_specific) const;
> ```
> 
> `GetDataDirNetPath()` would seem more readable than `GetDataDirPath(true)` at call sites, and harder to confuse when writing code. And since call sites are being updated in this PR anyway, implementing this would just require tweaking the scripted diffs.

I have attempted to do the change you propose. Please let me know if you find it better.

(I'm sorry about the needless rebase. I got confused by a test result.)

File: src/test/util/setup_common.cpp [hunk 1]
```diff
@@ -195,7 +195,7 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
     }
 
     m_node.addrman = std::make_unique<CAddrMan>();
-    m_node.banman = std::make_unique<BanMan>(m_args.GetDataDirPath() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
+    m_node.banman = std::make_unique<BanMan>(m_args.GetDataDirBase() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
```
Comment (ryanofsky) [inline src/test/util/setup_common.cpp:198]: In commit "Use `ArgsManager.GetDataDirBasePath()` and `ArgsManager.GetDataDirNetPath()` instead of `ArgsManager.GetDataDirPath(net_identifier)`" (80dec1717258fd74ae32274affed71daf2949e46)

IMO it would be a little better to keep the `GetDataDirPath` method and argumentless `GetDataDirPath()` calls unchanged in this commit, so the commit is smaller. Then the subsequent scripted-diff commit could do mechanical work of replacing `m_args.GetDataDirPath()` with `m_args.GetDataDirNet()` and `GetDataDirPath` with `GetDataDir`.

Also, it doesn't matter since it only affects tests, but replacing `GetDataDirPath()` with `GetDataDirNet()` instead of `GetDataDirBase()` would be a more obvious choice to keep behavior unchanged.

This is all fine, though. Just extra suggestions in case there's a reason to update.

**Review by ryanofsky — APPROVED** [review]
Code review ACK c8616c5c291c3c180460cf01ef3fcacd9194a2e4

File: src/util/system.h [hunk 2]
```diff
@@ -271,11 +271,18 @@ class ArgsManager
     /**
      * Get data directory path
      *
-     * @param net_specific Append network identifier to the returned path
      * @return Absolute path on success, otherwise an empty path when a non-directory path would be returned
      * @post Returned directory path is created unless it is empty
      */
-    const fs::path& GetDataDirPath(bool net_specific = true) const;
+    const fs::path& GetDataDirBase() const { return GetDataDir(false); }
```
Comment (MarcoFalke) [inline src/util/system.h:276]: 80dec1717258fd74ae32274affed71daf2949e46: The goal of this pull is to remove the `GetDataDir` function, but here you introduce more call sites?

Also, it could make sense to have this in a separate commit and move the other mechanical changes to a scripted diff

**Review by maflcko — CHANGES_REQUESTED** [review]

File: src/util/system.h (hunk 2 above)
Comment (kiminuo) [reply src/util/system.h:276]: > [80dec17](https://github.com/bitcoin/bitcoin/commit/80dec1717258fd74ae32274affed71daf2949e46): The goal of this pull is to remove the `GetDataDir` function, but here you introduce more call sites?

`GetDataDir(false)` actually calls `const fs::path& ArgsManager::GetDataDir(bool net_specific) const` (i.e. https://github.com/bitcoin/bitcoin/commit/80dec1717258fd74ae32274affed71daf2949e46#diff-a5795f7946088222aa59a662e15f5b5aee627bfe2311459e789d2a46783e37cbR455). But I can see that it is very actually very confusing. 

I'll try to go with https://github.com/bitcoin/bitcoin/pull/21850#discussion_r637092957 now and hopefully it will be nicer.

File: src/test/util/setup_common.cpp (hunk 1 above)
Comment (kiminuo) [reply src/test/util/setup_common.cpp:198]: I have moved the tests changes to a separate commit (https://github.com/bitcoin/bitcoin/pull/21850/commits/7f5774c09affebc8837a4d117aae198d6fd6157a). 

The PR differs in one trailing newline as an effect. So there is no code change, only git commit changes.

File: src/util/system.h [hunk 3]
```diff
@@ -268,6 +268,22 @@ class ArgsManager
...
+    const fs::path& GetDataDirBase() const { return GetDataDirPath(false); }
+
+    /**
+     * Get data directory path with appended network identifier
+     *
+     * @return Absolute path on success, otherwise an empty path when a non-directory path would be returned
+     * @post Returned directory path is created unless it is empty
+     */
+    const fs::path& GetDataDirNet() const { return GetDataDirPath(true); }
+
```
Comment (MarcoFalke) [inline src/util/system.h]: d5f1458ea3821cdf87954f012bd1606659cf818f: This is adding trailing whitespace. Also the commit title is wrong. `Path` should be removed.

File: src/util/system.cpp [hunk 4]
```diff
@@ -1361,7 +1361,7 @@ fs::path AbsPathForConfigVal(const fs::path& path, bool net_specific)
     if (path.is_absolute()) {
         return path;
     }
-    return fsbridge::AbsPathJoin(GetDataDir(net_specific), path);
+    return fsbridge::AbsPathJoin(net_specific ? gArgs.GetDataDirNet() : gArgs.GetDataDirBase(), path);
```
Comment (MarcoFalke) [inline src/util/system.cpp:1359]: 162b12cbc1051dbb8a25f271d0def1985f7cd73e: This can be squashed into d5f1458ea3821cdf87954f012bd1606659cf818f?

File: src/util/system.h [hunk 5]
```diff
@@ -453,6 +444,16 @@ class ArgsManager
     void LogArgs() const;
 
 private:
+
```
Comment (MarcoFalke) [inline src/util/system.h:446]: 849ea30efe96b29ec3023d10b077ffb4f7fa0f86: According to clang-format this newline should be removed

**Review by maflcko — APPROVED** [review]
review ACK 849ea30efe96b29ec3023d10b077ffb4f7fa0f86 🍇

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 849ea30efe96b29ec3023d10b077ffb4f7fa0f86 🍇
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgybwv/bU7GLhuy9uVt7h3E2faUAPQ2zkOGL9yViqsbXcZyqOUd5XYp8NQFt0Od
ADu7KAStRDEkqMCf0XMI9+Sx1loYVnzXACPvbf8b41+5yAerLOnmig08tDAIoLMc
YMuHK5rHjzrcLiwtrANmsW/jUR7l1pQY7QC5K0wWzSUX3XC69XZ0ZLHnJf2xJv+W
lr+xSFzbDJKcDE0PJQ3veswVGTHIkRw70QdkVj1Us3cX8i7cWADDNCsev9+DolBk
sojoZWDE58Gh/RKB5+dKyi1NpmsMjMJ4KN3l0Cgq6MU1Z8Q1ZiJ1UjwmQNIiGKqC
NNe352Op3nqbhIvXmciUgpTQcYveHmi8OAOv4hYwX2JMCmzM3PdPQ+TDgHuAY8El
PKH6mYTXfdqHow9x9QOpno67/Ei3ubzqmWsFnNTmG825lZEmb5YhviMbUvaEXWSy
6oum6G5mof5Mou4YI72d3cJbrMe0YCgA1P+SzK/jOEx2Hsp6rDw2+LIreBhtzffI
hmJY6Ucg
=7cLP
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `73e12c06b9e21bd03f9dc8f12311837a91583686ee46b861540ec6de544476d4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010873e12c06b9e21bd03f9dc8f12311837a91583686ee46b861540ec6de544476d4f010b14af1acfe0a2085a8e477413dae560f08fff010abc6209b99131c06f83aacb42528fd7308f10460a92f7ef008d1caca1f091ad40b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ae39a525261d934afd4a479219ba82fe08f10460a92f7ef008b15d41ca6ec63e480083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/util/system.h (hunk 5 above)
Comment (kiminuo) [reply src/util/system.h:446]: Would it make sense to add a clang-format linter to Cirrus or maybe to `lint-all.sh` to enforce proper code style?

Comment (MarcoFalke) [reply src/util/system.h:446]: There are cases where clang-format is "wrong", so we can't enforce this. Maybe suggestions could make sense, but they might be noisy.

File: src/util/system.cpp (hunk 4 above)
Comment (kiminuo) [reply src/util/system.cpp:1359]: Yes, it should. I can do it if another touch is needed.

Comment (MarcoFalke) [reply src/util/system.cpp:1359]: My ACK is the only one. I am happy to re-ACK if you decide to push.

File: src/util/system.h [hunk 6]
```diff
@@ -119,7 +118,7 @@ UniValue RunCommandParseJSON(const std::string& str_command, const std::string&
  * the datadir if they are not absolute.
  *
  * @param path The path to be conditionally prefixed with datadir.
- * @param net_specific Forwarded to GetDataDir().
+ * @param net_specific Forwarded to gArgs.GetDataDirNet().
```
Comment (hebasto) [inline src/util/system.h:121]: This comment is not true now.
Also suggesting to move this comment update into the _"Add `ArgsManager.GetDataDirBase()` and `ArgsManager.GetDataDirNet()` as an intended replacement for `ArgsManager.GetDataDirPath(net_identifier)`"_ commit (babd5ed0204b01793c8c96d079d5304efdcc36e5).

**Review by hebasto — COMMENTED** [review]
Approach ACK 457515aab1042d842ee2ac6752ebf577ae6dee2b.

File: src/util/system.h (hunk 6 above)
Comment (kiminuo) [reply src/util/system.h:121]: Right. I'm not sure about wording though. 

Maybe any of the following variants:

* `* @param net_specific Whether to append network identifier to base datadir first` or
* `* @param net_specific Use network specific datadir variant` or 
* `* @param net_specific Forwarded either to ArgsManager.GetDataDirNet() or ArgsManager.GetDataDirBase()`

**Comment by MarcoFalke** [comment]
ACK 457515aab1042d842ee2ac6752ebf577ae6dee2b 🚊

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 457515aab1042d842ee2ac6752ebf577ae6dee2b 🚊
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUibDgwAqnNrcU52+PTrpofeW2/FPyAMU3psifPGg3tyNYmI379eF9mbddfGk+Nv
heuP8JcCDj3JY1MPjqGsj6X5Bhof4zldeGHEpP/Zx+eW+IazU39RsdlCEv21X32n
+bCL7cnV4+6PJENrtKpJqxyET1Y6zguEjU+0O1kUeZdvYeLPP6K4SGgXKoYDxZl6
J3coLN7s0839Pn+qCNXFs/GEAjZR4j/RXXfbIUxx/L1bdBK5y/gAau5br/FMY+WS
ld465ZH1kOFdg80AyCxV4dY1B2QuIUw+zwE+8q7U7JVATAGOqQGkOBt/DKkd8+6U
2u7bbwkHieOba4XNfMXKvoApUmcD0pJTroSYo2P8Z9kSwVm7YCxGxBxP5Z2gC2qa
yKXQJkv07LUgMpXCIrfjky+hx4AVclLzlMYi800/8EAfrEDFxxPidUUEjOUFp8e2
cyNG8fkTd32Be41h0MDNidKQ+uLLv5dDin4cPDvHfNO7Q7qYuf/d7hSM3wnGeVwo
oVWPT/P3
=joGU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `cbcc6143812dabddaf1ceda66348759c62c6dfc96fa21f1fab33a29027c960f4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108cbcc6143812dabddaf1ceda66348759c62c6dfc96fa21f1fab33a29027c960f4f01044df937fa91b52e82d2de6f976becee508fff0102b4e17aa64f1fd1ec8d2a0bcca7cc54e08f02063b450ace5d5e3d293ff772806412add2024d1b74da7413383a7138b6f3ea1fe08f020b04e0618febe58e861f2094d2307e3d9c4b5df39ccb404dfb0afbfa258d4446408f10460ab452ef0081f33964b66cc967d0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01098d9633e15cd32592f64081ba71d6bed08f10460ab452df0080f7363e2a2bed3420083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010bfb90032df3cdd65adaf0cbcb80d41fb08f10460ab452ef0086dffb1f00d1d72b10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/util/system.h (hunk 6 above)
Comment (MarcoFalke) [reply src/util/system.h:121]: The `AbsPathForConfigVal` function should probably be made a method on `ArgsManager` anyway. The comment can be fixed up in the follow-up.

Happy to re-ACK if you push, though.

Comment (hebasto) [reply src/util/system.h:121]: The second one lgtm.

Comment (kiminuo) [reply src/util/system.h]: > The `AbsPathForConfigVal` function should probably be made a method on `ArgsManager` anyway. The comment can be fixed up in the follow-up.

Yes, it seems so.

> The second one lgtm.

Done. Thanks.

**Review by hebasto — APPROVED** [review]
ACK aca0e5dcdb174ef7e88b76910d6fffd633688235

**Review by maflcko — APPROVED** [review]
re-ACK aca0e5dcdb174ef7e88b76910d6fffd633688235 👃

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK aca0e5dcdb174ef7e88b76910d6fffd633688235 👃
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiN6Qv/RgP9uS/QKuYzESznN0Tb9TySTikTypK4lmJTJloWTVxqVMWSRaeey4wP
lgnhjfPjicCgooYTHZWb9T1qrQyFrjlRuyBkDANMY0zLvuLikxfTVEc+txoxvIib
cqYHxrf8sfKrv50JJZNpspRNBxgNrgl/0vdDNHisX3+rwm7/oRb6o5WXsWPa3xmv
3UKTnWZ75Z9XTRE5pqQKDtwGsidFay8GVCJPrq0NuB8o50nQBAByfUCD1thOdE5g
nNfunXEPM2hRwgy7/T8AL6fd69STFPoOuiSCWXX+5ASbWNUB8Mwl48o93DxZdIlK
/Sk3jk+q+zTl8nc6aOT3sKR3HK9P5XLUtY5tcmDH6kx9S89/IsCots2gEHL9R0Le
Wxu5hKBcttxMLH6dN/qLa8mK/Q5foDbFEft7Yv0XYYocI9Tc5APA5xkUyYng7Pb7
1FDXJC6Xal/+4aXHNwRLeXIaJgJ6LxQ4mq7xxlIGx5zUH53vIr88rL8B8uRBTn+5
DGxzYjRs
=z/lk
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f424bad1e3eed91a429f67534ae92bee38c1b1ae4c61eae9451b352c20c8a36b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f424bad1e3eed91a429f67534ae92bee38c1b1ae4c61eae9451b352c20c8a36bf010826a92f14b0de1559bed2ba61b21fb8008fff010106a9b55412d7edddd3ef5286b88f0ad08f10460ab6c4ff0081d899b6b38a854950083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01041654d94d63f85ae92344e8889a5dca108f10460ab6c50f008984f91c279e606060083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010a455be6d84a038540cba9682b2a1c65c08f10460ab6c4ef00838f41f1265f969100083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
